### ISSUE TITLE
Investigate whether we can compat-transpiler lookbehind assertions

### ISSUE THREAD

**Issue by DmitrySoshnikov** [DmitrySoshnikov/regexp-tree#66]
The [lookbehind assertions](https://github.com/DmitrySoshnikov/regexp-tree#lookbehind-assertions) (stage 3 [proposal](https://tc39.github.io/proposal-regexp-lookbehind/)) allow matching a pattern only if it's preceded (or not preceded for negative) by another pattern.

We should investigate, whether it's possible (in any approximation) to compat-transpile them to ES2018->ES2015.

Matches `y` only if it's preceded by `x`:

```js
/(?<=x)y/
```

One of the approaches is to use non-capturing group:

```js
/(?:x)y/
```

We're talking about covering `exec`, and other methods, since for just `replace` method it'd be possible to emulate as:

```js
/(?<=x)y/
```

```js
// <match>x
'xyx'.replace(/(x)?y/g, ($0, $1) => $1 ? '<match>' : $0);
```

But we need a more-less acceptable solution for transformation.

**Comment by hg42** [comment]
I think these assertions cannot be transformed into current javascript RegExp.

That's because they are *assertions*, which means they don't consume anything.

E.g.
```
/(..)(?<=x)y/
```
would capture the x and the preceding character in the first group.
You can expressed weird patterns with it.
I always tried to avoid them because it's an expensive operation and not really easy to understand all consequences.
So everytime I wanted to have them in my programming live, I tried to find alternatives and always succeeded.

I'm sure they are necessary. I just didn't find any use-case, yet.

**Comment by hg42** [comment]
to be more clear:
look behind is easy for easy cases.

Mainly it's match isn't included in or excluded from the match results.
So it effects result[0] and doesn't steal character from the expression in front of it.

But because it anchors the following expressions it can be tricky inside complex nested expressions.
It's similar to the effects of a `\b` assertion or the famous `$^`-thing Dmitry wrote about.

**Comment by hg42** [comment]
another example (or test case), where the assertion is for a longer string than the captured group:
```
/(..)(?<=bcdx)y/
```
applied to
```
abcdxyz
```
returns `'dx'` for the first group (result[1]) and `'dxy'` for the whole match (result[0]).

You can see assertion as a second layer tested in parallel to the rest of the expression.
For a moment I thought, it should be possible to check the assertions in a second pass, checking if the assertions are all met.
But the tricky part is the anchoring.
So, another idea was to find the expressions surrounded by assertions using something like
```
(look_behind_assertion)(expression)(look-forward-assertion)
```
and split the text. Then the other parts of the text and the regexp could be processed separately.
But think about this in case of complex nested expressions. You'll certainly end up with an engine having similar complexity like a regexp engine.

It might get even trickier:
```
/(?<=xxx|N)(?<a>x)/g
```
matches like this:
```
abcxxxxxxNxxxxxx
------aaa-a--aaa
```
Whereas
```
/(?:xxx|N)(?<a>x)/g
```
matches like this:
```
abcxxxxxxNxxxxxx
------a---a---a-
```
You see, the behaviour is very different.
The `'xxx'` are matched but not consumed by the look behind expression, which means they can match again.

Note, I use `/.../g` here but it's the same for any kind of repetition.

And that's not the end. You can have expressions, where the same string is matched by several look behind assertions in parallel. There are priorities deciding which one will match
E.g.:
```
/(?<=xxx)(?<a>x)|(?<=....)(?<b>x)/g
```
with the same text from above gives
```
abcxxxxxxNxxxxxx
----bbaaa-bbbaaa
```
but if I swap the two parts:
```
/(?<=....)(?<b>x)|(?<=xxx)(?<a>x)/g
```
they match like this:
```
abcxxxxxxNxxxxxx
----bbbbb-bbbbbb
```
ok, let's stop here :-)

Conclusion: it's not a simple transformation but very different behaviour.

**Comment by DmitrySoshnikov** [comment]
@hg42, thanks for the detailed analysis (as always :)), good points. Yeah, I think we should postpone with this issue for now, since it's not trivial, and the impact probably is not that big to worry about it.

**Comment by DmitrySoshnikov** [comment]
Closing this for now.

**Comment by davisjam** [comment]
## Summary

Reviving this based on the question in #201. I think it's possible to *simulate* lookbehind assertions using traditional regexes, though I'd welcome a second opinion.

## Simple approach won't work

Because these are zero-width assertions, they can't be transpiled with a trivial syntactic transformation as was initially proposed. For example,

- `/^x(?<=x)y/` would match the string `xy` but not `xxy`.
- `/^x(?:x)y/` would match the string `xxy` but not `xy`.

## Intersection?

However, lookbehind assertions can in many cases* be simulated using a regex intersection algorithm. A lookbehind assertion of the form `/Sub-pattern 1 (?<= Sub-pattern 2)/` will match all of the strings in the *intersection* of `/Sub-pattern 1/` and `/.*Sub-pattern 2$/` (i.e. the strings that match sub-pattern 1 and end in sub-pattern 2). Since regular expressions are closed under intersection**, applying a regex/NFA intersection algorithm to the two will produce another NFA/regex***.

By the same argument, a lookahead assertion can be simulated by an intersection of the lookahead sub-pattern and the *subsequent* sub-pattern.

## Tricky bits

Regex intersection algorithms are concerned solely with being language-preserving, and ignore applied details like capture groups. So preserving behavior beyond matching, e.g. capture groups, may be difficult.

## Footnotes

*AFAIK These algorithms have only been developed for "true" regular expressions, e.g. `/a|b*c/`-type things but not more exotic features like backreferences.

**Again, only for "true" regular expressions, though some recent work by Campeanu and others suggests it may be true of some more advanced features as well.

***The intersection algorithms I know of operate on NFAs, not regexes, and the only algorithms I know of for concocting a readable regex from an NFA may result in a "low-quality" regex (see, e.g., the flow algorithm, which yields a potentially exponentially-long and repetitive-looking regex). But perhaps there are intersection algorithms based on algebraic manipulation of the regex itself, e.g. in conjunction with Brzozowski's "derivatives" approach, which might yield a "higher-quality" regex?

**Comment by DmitrySoshnikov** [comment]
@davisjam, thanks, all good points. As you also correctly mention, this might be possible to some extent, however JS regular expressions go beyond simple NFA implementations (with capturing groups, back-references, etc).

It's a good theoretical field, however I didn't find yet to translate lookbehinds for all the cases in JS. _Probably_, we could allow translation this small subset, as long as it produces still performant regexes, and is practical.
